processing/src/main/java/org/apache/carbondata/processing/newflow/steps/writer/DataWriterProcessorStepImpl.java

```diff
@@ -0,0 +1,360 @@
+    }
+
+    CarbonTable carbonTable = CarbonMetadata.getInstance().getCarbonTable(
+        tableIdentifier.getDatabaseName() + CarbonCommonConstants.UNDERSCORE + tableIdentifier
+            .getTableName());
+    wrapperColumnSchema = CarbonUtil
+        .getColumnSchemaList(carbonTable.getDimensionByTableName(tableIdentifier.getTableName()),
+            carbonTable.getMeasureByTableName(tableIdentifier.getTableName()));
+    colCardinality = CarbonUtil.getFormattedCardinality(dimLensWithComplex, wrapperColumnSchema);
+    segmentProperties = new SegmentProperties(wrapperColumnSchema, colCardinality);
+    // Actual primitive dimension used to generate start & end key
+
+    keyGenerator = segmentProperties.getDimensionKeyGenerator();
+
+    //To Set MDKey Index of each primitive type in complex type
+    int surrIndex = simpleDimsCount;
+    Iterator<Map.Entry<String, GenericDataType>> complexMap =
+        CarbonDataProcessorUtil.getComplexTypesMap(configuration.getDataFields()).entrySet()
+            .iterator();
+    complexIndexMap = new HashMap<Integer, GenericDataType>(complexDimensionCount);
+    while (complexMap.hasNext()) {
+      Map.Entry<String, GenericDataType> complexDataType = complexMap.next();
+      complexDataType.getValue().setOutputArrayIndex(0);
+      complexIndexMap.put(simpleDimsCount, complexDataType.getValue());
+      simpleDimsCount++;
+      List<GenericDataType> primitiveTypes = new ArrayList<GenericDataType>();
+      complexDataType.getValue().getAllPrimitiveChildren(primitiveTypes);
+      for (GenericDataType eachPrimitive : primitiveTypes) {
+        eachPrimitive.setSurrogateIndex(surrIndex++);
+      }
+    }
+  }
+
+  private void initDataHandler() {
+    int simpleDimsCount = dimensionCount - complexDimensionCount;
+    int[] simpleDimsLen = new int[simpleDimsCount];
+    for (int i = 0; i < simpleDimsCount; i++) {
+      simpleDimsLen[i] = dimLens[i];
+    }
+    CarbonDataFileAttributes carbonDataFileAttributes =
+        new CarbonDataFileAttributes(Integer.parseInt(configuration.getTaskNo()),
+            (String) configuration.getDataLoadProperty(DataLoadProcessorConstants.FACT_TIME_STAMP));
+    String carbonDataDirectoryPath = getCarbonDataFolderLocation();
+    CarbonFactDataHandlerModel carbonFactDataHandlerModel = getCarbonFactDataHandlerModel();
+    carbonFactDataHandlerModel.setPrimitiveDimLens(simpleDimsLen);
+    carbonFactDataHandlerModel.setCarbonDataFileAttributes(carbonDataFileAttributes);
+    carbonFactDataHandlerModel.setCarbonDataDirectoryPath(carbonDataDirectoryPath);
+    carbonFactDataHandlerModel.setIsUseInvertedIndex(isUseInvertedIndex);
+    if (noDictionaryCount > 0 || complexDimensionCount > 0) {
+      carbonFactDataHandlerModel.setMdKeyIndex(measureCount + 1);
+    } else {
+      carbonFactDataHandlerModel.setMdKeyIndex(measureCount);
+    }
+    dataHandler = CarbonFactHandlerFactory.createCarbonFactHandler(carbonFactDataHandlerModel,
```

Comment:
I think we should make it work then refactory the `CarbonFactHandler` later. We should decouple the producer-consumer and actual writing logic. There should be abstraction like `CarbonFileWriter` and `CarbonBlockletWriter`, and it can be used to write single carbon data file by java application.